### ISSUE TITLE
GUI: A base dialog for general-purpose dialogs with multiple settings

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#4620]
**Reported by nvdakor on 2014-11-12 17:50**
Hi,
Recent NVDA code added several general-purpose dialogs such as the new exit dialog and update check dialog. As opposed to specialized dialogs such as find dialog in browse mode, these dialogs have a common weakness: multiple instance possibility, mitigated with use the weakrefs and an instance variable. Another weakness is the problem of window placement: manual intervention is needed to center them and to provide needed sizers.
To provide a common base case where NvDA core or add-ons may display general-purpose dialogs that are not settings dialogs, I propose a new base dialog class in GUI subsystem as follows:
- Dialog: This new dialog class in GUI will represent a new general-purpose dialog with much needed attributes filled in, such as centering, sizers, handlers for onOK, onCancel and options sizer. It'll borrow somewhat from Settings Dialog class for now unless the devs elect to base the Settings Dialog from this new dialog class (with settings dialog inheriting from the new base dialog). The attributes for this dialog will be its title, the main box sizer, customizable OK and Cancel buttons and events for them, a general-purpose sizer, layout attributes such as centering and an instance variable to make sure this dialog (or its children) appear only once.
- Dialogs in NVDA core will be modified to base itself from the new base dialog. For example, the new ExitDialog will be written as:

```
class ExitDialog(gui.Dialog):
""" code """
```

All that'll be added will be the new options for exiting. The dialogs to be modified include exit dialog, donations, update check and alerts, cursor manager's find dialog, welcome dialog, installer and possibly others. I excluded app module and object specific dialogs such as comments dialog in Excel, as they need to be placed near the cell where the comments will appear. In the end, Settings Dialogs will also be rewritten to inherit from the new base dialog if the devs elect to do so.
- In add-ons, whenever an add-on needs to present a dialog that is not a settings dialog, have them inherit from gui.Dialog class. Even when they add settings dialogs, they are ultimately deriving from the base dialog.
  Benefits of this change include ability to change the appearance of vast majority of dialogs by modifying the new base class, as well as to prepare for future WXPython changes for dialog code. Drawbacks include implementation strategies, tuning existing dialogs to use the new scheme and overlap with settings dialog code.
  Thanks.

**Comment by feerrenrut** [comment]
A common base class could perhaps override the `Show` and `ShowModal` methods in order to ensure that `prePopup` / `postPopup` happen, see [comment](https://github.com/nvaccess/nvda/pull/6355#discussion_r167452784) on "Introduce the ability to postpone a downloaded update" #6355

As discussed in the linked comment, it's not certain that newer operating systems or versions of wx require this. Linking / suggesting for completeness sake.

**Comment by josephsl** [comment]
Hi,

2024 update: I think there is an issue in 2025.1 milestone to do something about dialogs. We could combine this issue with that one or leave the issues separate.

Thanks.

**Comment by Adriani90** [comment]
cc: @seanbudd for triaging

**Comment by seanbudd** [comment]
@josephsl - #13007 is a fix for Message dialogs in particular, I'm not sure if this solves the same problems. I don't know what the original reporters goals were, but I think it would be worth checking #13007 and thinking if this solution covers what is needed.
@Adriani90 - I don't think this can be triaged in it's current state as the original issue's goals are unclear, and the discussion has seemed to migrated away from the initial proposal. I would probably encourage a new issue or a comment on #13007 depending on what is wanted here.

Ccing @SaschaCowley who's working on #13007

**Comment by josephsl** [comment]
Hi,

I was the original reporter...

The idea (back then) was to offer a common base class with an ability to add additional controls, mostly settings. Now that we have a basic settings panel class, part of that proposal is implemented. However, there are add-ons that provide settings in their dialogs, so I figured it would be good to offer a more consistent experience such as centering these dialogs without each dialog calling wx.CenterOnScreen() function (I wish add-ons could put their settings inside NVDA settings panels, but it's up to the add-on author to take advantage of this arrangement; there are add-ons requiring a more custom settings experience such as their own settings panels collection (StationPlaylist add-on is a good example of this)).

Thanks.

**Comment by seanbudd** [comment]
@josephsl - could you perhaps update the issue with the template and what you think is remaining?

**Comment by josephsl** [comment]
Hi, at this point I think it would be best to create an updated issue from scratch. Thanks.

**Comment by SaschaCowley** [comment]
In light of @josephsl's [comment](https://github.com/nvaccess/nvda/issues/4620#issuecomment-2387628768), I'm closing this.
